File: src/303_range_sum_query_immutable/range_sum_query_immutable.py
from typing import List


class NumArray:
    def __init__(self, nums: List[int]):
        self.prefSum = []
        self.prefSum.append(nums[0])

        for i in range(1, len(nums)):
            self.prefSum.append(self.prefSum[i - 1] + nums[i])

    def sumRange(self, left: int, right: int) -> int:
        sum = self.prefSum[right]
        if left > 0:
            sum -= self.prefSum[left - 1]

        return sum


# Test scaffolding
if __name__ == "__main__":
    numArray = NumArray([-2, 0, 3, -5, 2, -1])
    tests = [
        ((0, 2), 1),
        ((2, 5), -1),
        ((0, 5), -3),
        ((3, 3), -5),
        ((1, 1), 0),
    ]
    for (left, right), expected in tests:
        actual = numArray.sumRange(left, right)
        print(f"{actual} | {expected}")

File: src/303_range_sum_query_immutable/range_sum_query_immutable.cpp
#include <iostream>
#include <vector>
using namespace std;

class NumArray {
    vector<int> prefix;

   public:
    NumArray(vector<int>& nums) {
        prefix.push_back(nums[0]);

        for (int i = 1; i < nums.size(); i++) {
            prefix.push_back(prefix[i - 1] + nums[i]);
        }
    }
    int sumRange(int left, int right) {
        int sum = prefix[right];
        if (left > 0) sum -= prefix[left - 1];

        return sum;
    }
};

int main() {
    vector<int> nums = {-2, 0, 3, -5, 2, -1};
    NumArray numArray(nums);
    vector<pair<pair<int, int>, int>> tests = {
        {{0, 2}, 1}, {{2, 5}, -1}, {{0, 5}, -3}, {{3, 3}, -5}, {{1, 1}, 0}};
    for (auto& [query, expected] : tests) {
        int actual = numArray.sumRange(query.first, query.second);
        cout << actual << " | " << expected << endl;
    }
    return 0;
}
